### ISSUE TITLE
Expanding documentation with error handling examples

### ISSUE THREAD

**Issue by shoooe** [graphql-dotnet/graphql-dotnet#495]
The documentation is not very clear with regards to how to:
- Create custom errors
- Throw these custom errors
- Customize the rendering of these errors to JSON (for example how to add fields to the `errors` response object)

Is there any example out there that showcases these?

**Comment by joemcbride** [comment]
Agreed, that should be added to the docs.

>Create custom errors
>Throw these custom errors

At present, you can throw any error you would like and it should be caught and wrapped in an `ExecutionError`.  Otherwise you can create an `ExecutionError` yourself and add it to the context which will get bubbled up.

```csharp
Field<DroidType>(
  "hero",
  resolve: context =>
  {
    context.Errors.Add(new ExecutionError("Custom error"));
    return null;
  }
);
```

> Customize the rendering of these errors to JSON (for example how to add fields to the errors response object)

This is not currently possible.  This is the code that handles writing errors at present.

https://github.com/graphql-dotnet/graphql-dotnet/blob/c733e48fe38fb22d2360af35379cfbd6bbca8a36/src/GraphQL/Execution/ExecutionResultJsonConverter.cs#L36

**Comment by Jefffrey** [comment]
I see that ExecutionError has a Data property. Can this property be set somehow?

**Comment by joemcbride** [comment]
`ExecutionError` just inherits from `Exception`.

https://github.com/graphql-dotnet/graphql-dotnet/blob/c733e48fe38fb22d2360af35379cfbd6bbca8a36/src/GraphQL/Execution/ExecutionError.cs#L8

From my understanding, should be able to add values to it like this (since it is just an `IDictionary`):

```csharp
var exc = new ExecutionError("Some error");
exc.Data.Add("avalue", "value");
```

Though that won't do anything unless the `ExecutionResultJsonConverter` is updated to write json from that property.

**Comment by alexandercarls** [comment]
I am especially interested for error handling as a cross-cutting concern.
For example when my business logic within the resolver throws a `CustomDomainException`.

I would like to have the opportunity e.g. with a Middleware to catch this specific exception and translate it to a GraphQL specific `ExecutionError`.

**Comment by benmccallum** [comment]
Did you ever find a way to do that @alexandercarls? I think I have a use case for that too.

**Comment by BenjaBobs** [comment]
You could write a base resolver that all your resolvers inherit from, that has a try..catch to catch that specific exception.

**Comment by benmccallum** [comment]
But aren't resolvers just `Func`s, so they aren't classes that can inherit from another. You mean a Func that wraps (and internally executes) your actual resolver Func?

**Comment by benmccallum** [comment]
Started looking at `IFieldResolver` and although it would allow me to create my own abstract base `MyFieldResolver : IFieldResolver`, I'm using async await, so it looks like I'd need to implement/inherit an `AsyncFieldResolver<TIn, TOut>`. 

Starts getting weird quick though. Inheriting `AsyncFieldResolver<TIn, TOut>` means I need to pass my resolver in the call to `base` constructor. Not really a problem but kinda funny looking. 

```csharp
public class BookingPayingMethodResolver : AsyncFieldResolver<BookingDto, BookingPayingMethodDto>
{
    public BookingPayingMethodResolver(IPaymentClient paymentClient, IDataLoaderContextAccessor accessor)
        : base(async context =>
        {
            var loaderKey = $"{nameof(IPaymentClient)}::{nameof(IPaymentClient.GetBookingPayingMethodsAsync)}";
            var loader = accessor.Context.GetOrAddBatchLoader<int, BookingPayingMethodDto>(loaderKey, async (ids) =>
            {
                var request = new GetBookingPayingMethodsRequest(1, ids.Count(), ids.ToHashSet());
                var response = await paymentClient.GetBookingPayingMethodsAsync(request);
                response.CheckForError();
                return response.Items.ToDictionary(bpm => bpm.BookingId);
            });
            return await loader.LoadAsync(context.Source.BookingID);
        })
    { }
}
```

Not inheriting but simply taking inspiration from `AsyncFieldResolver<TIn, TOut>` I get close but I don't really understand how it works. It does a call to `.As<TIn>` on the `ResolveFieldContext` so that it can be passed to the resolver func as `ResolveFieldContext<TIn>`, but that extension method for me just casts it to `TIn`. Am I missing an implicit cast or getting the wrong extension method? 

```csharp
public class BookingPayingMethodResolver2 : IFieldResolver<Task<BookingPayingMethodDto>>, IFieldResolver
{
    private readonly IPaymentClient _paymentClient;
    private readonly IDataLoaderContextAccessor _accessor;

    public BookingPayingMethodResolver2(IPaymentClient paymentClient, IDataLoaderContextAccessor accessor)
        : base()
    {
        _paymentClient = paymentClient;
        _accessor = accessor;
    }

    public Task<BookingPayingMethodDto> Resolve(ResolveFieldContext context)
    {
        var contextTyped = context.As<BookingDto>();
            
        var loaderKey = $"{nameof(IPaymentClient)}::{nameof(IPaymentClient.GetBookingPayingMethodsAsync)}";
        var loader = _accessor.Context.GetOrAddBatchLoader<int, BookingPayingMethodDto>(loaderKey, async (ids) =>
        {
            var request = new GetBookingPayingMethodsRequest(1, ids.Count(), ids.ToHashSet());
            var response = await _paymentClient.GetBookingPayingMethodsAsync(request);
            response.CheckForError();
            return response.Items.ToDictionary(bpm => bpm.BookingId);
        });

        // FAILS to compile here as context is of type `BookingDto` instead of ResolveFieldContext<BookingDto>`
        return await loader.LoadAsync(context.Source.BookingID);
    }

    object IFieldResolver.Resolve(ResolveFieldContext context)
    {
        return Resolve(context);
    }
}
```

I'm keen to get something nice here as the other benefit is the resolver is in another *.cs file rather than having all your resolvers defined in your graph type, which could get pretty unwieldly over time.

**Comment by BenjaBobs** [comment]
Well, the whole inheritance chain of contexts and resolvers is a bit confusing, I was actually hoping that there would be a discussion to change it at some point. I haven't tried to implement the scenario you're describing, so I don't currently have an overview of the process.

**Comment by alexandercarls** [comment]
I made a similar detour and couldn't get it to work nicely without forking the project.
For this current project I decided that our domain exceptions just inherit from `ExecutionError`.

I only had one idea left which I didn't explore yet.
In the GraphQL Middleware there might be a spot where I could translate domain exceptions in to `ExecutionError`s.

```csharp
        private async Task ExecuteAsync(HttpContext context, ISchema schema) {
            var request = Deserialize<GraphQLRequest>(context.Request.Body);
            var dataLoaderDocumentListener = serviceProvider.GetRequiredService<DataLoaderDocumentListener>();

            var result = await executer.ExecuteAsync(opt => {
                opt.Schema = schema;
                opt.Query = request.Query;
                opt.OperationName = request.OperationName;
                opt.Inputs = request.Variables.ToInputs();
                opt.UserContext = settings.BuildUserContext?.Invoke(context);
                opt.Listeners.Add(dataLoaderDocumentListener);
                opt.ThrowOnUnhandledException = true;
                opt.ExposeExceptions = env.IsDevelopment();
            });
            // --> Iterate result.Errors and replace them here.
            await WriteResponseAsync(context, result);
        }
```

As for your resolver structure. I used a mediator pattern with some rather large abstractions. My resolvers are only sending a message object with some props and the mediator is invoking the correct handler. That way my logic is agnostic of GraphQL (aside from the domain exceptions)

**Comment by benmccallum** [comment]
Good point about the middleware. In the end it turned out the thing I thought would throw exceptions and I'd want to catch and transform isn't, it just returns an empty response object essentially with a status flag on it. So I've just got a extension method on it's type `.ThrowIfNotValid()` or similar and it throws the exception I want that'll flow through into the GrpahQL errors. 

It's great that more people are getting real world experience using all this stuff to the point where we can iterate on the core framework / server and add in the extension points. 

I thought about chucking in MediatR, but I'm essentially wrapping back-end Web APIs with my GraphQL server, so that's about as big as my resolvers *should* get and essentially I end up doing the same thing as you (building up a request/message object with some props and passing it to the relevant Web API client. So for now I'm hesitant to chuck in a mediator pattern in there as it's just another abstraction, but I guess I'll see how unit testable this all is and go from there. Potentially I may need the mediator abstraction when it comes to mutations that need to call out to several web APIs... I'm not sure if building GraphQL on existing microservices/web apis is a blessing or a curse yet.

Thanks for the help!

**Comment by joemcbride** [comment]
Another option is to use field middleware to catch and translate domain exceptions.

https://graphql-dotnet.github.io/docs/getting-started/field-middleware

**Comment by alexandercarls** [comment]
@joemcbride Maybe I am fundamentally miss understanding async/await or the middleware.
I tried that before and the catch block never gets executed.

I do agree, that would be the best case scenario for translating domain exceptiong as a cross-cutting concern.

```csharp
public class ErrorHandlingFieldsMiddleware {
    public async Task<object> Resolve(
        ResolveFieldContext context,
        FieldMiddlewareDelegate next) {
        try {
            return await next(context);
        }
        catch (Exception e) {
            Console.WriteLine(e);
            throw;
        }
    }
}
```

**Comment by ffalsino** [comment]
Hello @joemcbride 
Can you please post a simple code example about how to use
https://graphql-dotnet.github.io/docs/getting-started/field-middleware  ?

Thank you so much

**Comment by alexandercarls** [comment]
FYI The middleware now properly works with this Pull Request: https://github.com/graphql-dotnet/graphql-dotnet/pull/865/files#diff-b4c19b6d66951c1c5d547222ae0d874bR22 and can be used to translate the exceptions.

**Comment by sungam3r** [comment]
It would be nice if someone contribute.

**Comment by benmccallum** [comment]
I guess it's just a matter of documenting an example of a custom middleware, similar to  `InstrumentFieldsMiddleware` or perhaps also like GraphQL.Validation's middleware that catches the FluentValidation ValidationException, and demonstrates what you can do. 

Does anyone wanna have a crack and mind if we assign this to them?

**Comment by denisedelbando** [comment]
Hello guys. Here is a sample that got it working for me.. 
So our entire site has a ServiceException that returns business error codes, proper http status and custom error message. this is how we are catching it. 

```
public async Task<object> Resolve(
            ResolveFieldContext context,
            FieldMiddlewareDelegate next)
        {
            object result = null;
            try
            {
                result = await next(context);
                return result;
            }
            catch (ServiceException serviceException)
            {
                var metadata = new Dictionary<string, object>
                {
                    
                    {"typeName", context.ParentType.Name},
                    {"source", context.FieldName},
                    {"innerException", serviceException.InnerException != null ? serviceException.InnerException.Message : "" },
                    {"stackTrace", serviceException.StackTrace },
                    {"message", serviceException.Message },
                    {"httpCode", serviceException.StatusCode }

                };
                var exception = new ExecutionError(serviceException.Message, metadata);
                exception.Code = serviceException.Code.ToString();
                exception.Path = context.Path;
                exception.Source = context.Source.ToString();
                context.Errors.Add(exception);
                return result;
            }
        }
```
Sample result
![sc-20200421004006-chrome](https://user-images.githubusercontent.com/3780000/79826846-c732f900-836a-11ea-83b0-a6ee9256a0c3.png)

**Comment by Shane32** [comment]
All of the issues here have been fixed in 3.0:

> The documentation is not very clear with regards to how to:
>
> * Create custom errors
> * Throw these custom errors

Error handling documentation has been updated in #1812 

> * Customize the rendering of these errors to JSON (for example how to add fields to the `errors` response object)

This is now somewhat possible with #1656 by writing your own implementation of `IErrorInfoProvider`.  With the `IErrorInfoProvider` interface, you can fully customize the `message` and `extensions` properties that are returned in the error object, but not the `path` and `locations` properties.  You can also add arbitrary fields to the `extensions` property of the error object, as per the spec that is the only spot to add arbitrary fields.  I suggest deriving from `ErrorInfoProvider` to write a custom implementation.  An example is shown in the documentation (added in PR #1816).

You can of course create your own serializer if you require an out-of-spec implementation.

> I am especially interested for error handling as a cross-cutting concern.
>
> For example when my business logic within the resolver throws a CustomDomainException.
>
> I would like to have the opportunity e.g. with a Middleware to catch this specific exception and translate it to a GraphQL specific ExecutionError.

This is easy with the `UnhandledExceptionDelegate` (added in #1164) like this:

```csharp
var executionOptions = new ExecutionOptions()
executionOptions.UnhandledExceptionDelegate = context =>
{
    if (context.Exception is ApplicationException)
    {
        context.Exception = new ExecutionError("An application exception has occurred");
    }
};
```

Also note that with the changes in #1656, the `UnhandledExceptionDelegate` only executes when a non-`ExecutionError` is thrown, as `ExecutionError`s are by definition "handled" errors.

Although it is still possible, there should be no need to write middleware for the sole purpose of catching, logging, or modifying exceptions.
